Pull Request: Feature/dangle

roboshoes opened the pull request:
This pull request adds a fuzzy way to calculate drift and rotation on crash. So far no physics engines have been used. Also `movable` gets an extra attribute called `torque`which represents the power released in ever curve segment. 

So far the dangeling of the car is not a hundred percent realistic but that might be a question of balancing. I though it might be nice to add it into the development branch for now to have it tested. 

_This entire feature is very easily removed, in case we hate it ..._

roboshoes committed:
adds fuzzy torque in movable

roboshoes committed:
WIP adds elastic target point behind car

roboshoes committed:
adds fuzzy dangle

roboshoes committed:
fixes tests and refactors

stravid commented:
I suggest we don't merge this PR but keep it to discuss the current status of the feature.

My thoughts:
- I love the way the car is drifting out if you go to fast. Great work!
- The dangling in general is to fast / jumpy.

roboshoes commented:
I fully agree. The dangling must be speed dependent (which it kind of is) but as soon as the car breaks down the dangling mussst be way less!

I keep working on it. Feedback appriciated!

roboshoes committed:
inertia is calculated based on speed

bmedicke commented:
moar dangle!

roboshoes committed:
removes drift when to slow and sets max drift if relative speed is greater than 0.6

roboshoes commented:
OK - I did a little fine tuning. Please, everybody check out the branch and do a little testing.

roboshoes committed:
sets the center of the car

stravid commented:
It's getting a lot better, nice job! But I still think we can do better, make it feel even more awesome :-)

bmedicke commented:
We have to decide if we want to simulate a real car or a slot car. I think it looks fine for a slot car. Maybe the car dangles a bit too fast, but the general behavior (how far the back moves) is fine.

roboshoes committed:
balanced dangle

tombu commented on app/assets/javascripts/slotcars/play/views/car_view.js.coffee:
```diff
@@ -17,28 +17,25 @@
 
     # if @offset?
     #   (jQuery @$()).css 'top', @offset.top
     #   (jQuery @$()).css 'left', @offset.left
```
I know this is nothing that belongs to your feature but as we don´t need `offset` any longer you could do me a favour an remove those 3 lines, couldn´t you? :wink:

tombu commented:
Please, please, please write some comments! I can´t understand the (= your) logic in `moveTo()` ...

roboshoes commented:
True - that can be removed just like the exhaust in general. it wouldn't work right now anyways. So it needs to be taken out.

DominikGuzei commented on app/assets/javascripts/slotcars/shared/lib/movable.js.coffee:
```diff
@@ -6,13 +6,78 @@ Vector = helpers.math.Vector
+  # constants for bounce calculation
+  INERTIA: 0.8
+  OFFSET: 0.1
+  K: 0.16
```
what does K mean?

Please add comments to describe how these constants influence the behaviour of the car

DominikGuzei commented on app/assets/javascripts/slotcars/shared/lib/movable.js.coffee:
```diff
@@ -6,13 +6,78 @@ Vector = helpers.math.Vector
-    @set 'rotation', direction.clockwiseAngle()
+    if @get 'isCrashing'
+      @set 'rotation', (@rotation + @torque)
+      @torque *= 0.94
```
no magic numbers -> move 0.94 to its own constant at the top of file with a descriptive name

DominikGuzei commented on app/assets/javascripts/slotcars/shared/lib/movable.js.coffee:
```diff
@@ -6,13 +6,78 @@ Vector = helpers.math.Vector
 (namespace 'slotcars.shared.lib').Movable = Ember.Mixin.create
```
Why is the code for dangle in Movable? Please move the whole dangle logic into an e.g `Dangleable` mixin to keep the file size small and the code testable!

DominikGuzei commented:
Please move the whole dangle logic into its own mixin and add comments + comprehensive tests for the logic. It's ok to try things out and perfect everything without writing tests but before we merge that into the project we need good understandable code that is well tested and can be changed later on without shooting yourself ;-)

roboshoes commented:
Unfortunatly you are completly wrong and I am right (as most of the time)!

There is a reason why this _logic_ is all in the movable. The dangeling only affects the rotation of the car which is definitely a part of movable. (Even by name: it moves the car). If you would have read my code closely you would have realized that I only work within this one function `moveTo`. Your argument to write it in form of inheritance (or for the sake of that as a mixin) and override `moveTo` would be completely senseless and collide with almost every coding law because it would create pointless unused code that **MIGHT ONE DAY IN 17 YEARS** be used by somebody who forks the project and creates another game.
In our specific case. I rewrote the calculation of the angle of the car (which by the way also fixes the jumpstart mess). I can not reuse the old `moveTo` function so overriding it would just through it away.
Also - the only thing, that is public, that I added is the `torque` which is tested by the way. You all would agree that you would not test lines within a function, or would you?

So in short I believe you can either change `Movable` or create a `Dangable`and throw `Movable` away which in other words would mean renaming it.

PS: I will rename most of the variables.
